### PR TITLE
VK_EXT_graphics_pipeline_library

### PR DESCRIPTION
**Pull request by ncesario-lunarg**

### CODE REVIEW

**Comment by ncesario-lunarg** [comment]
This PR is broken up into the following for (hopefully?) easier review:
- [ ]  https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/3958 Initial refactor prep
- [ ] https://github.com/ncesario-lunarg/Vulkan-ValidationLayers/pull/1 Make a number of PIPELINE_STATE fields private
- [ ] https://github.com/ncesario-lunarg/Vulkan-ValidationLayers/pull/6 Initial support for VK_EXT_graphics_pipeline_library
- [ ] https://github.com/ncesario-lunarg/Vulkan-ValidationLayers/pull/3 Fixes for CTS and other apps
- [ ] https://github.com/ncesario-lunarg/Vulkan-ValidationLayers/pull/4 VK_EXT_graphics_pipeline_library VUIDs (incomplete)
- [ ] https://github.com/ncesario-lunarg/Vulkan-ValidationLayers/pull/5 More fixes (including CI build fixes)

**Comment by ncesario-lunarg** [comment]
> Those are on your fork. Do you want them reviewed over there or was that unintentional?

On my fork should be good. I wasn't sure what Github would do, since each PR is "chained" together, and part of the chain is in the main repo. Plus I didn't want to flood our CI with changes that probably weren't going to pass all of CI.

**Comment by jeremyg-lunarg** [comment]
> > Those are on your fork. Do you want them reviewed over there or was that unintentional?
> 
> On my fork should be good. I wasn't sure what Github would do, since each PR is "chained" together, and part of the chain is in the main repo. Plus I didn't want to flood our CI with changes that probably weren't going to pass all of CI.

Ok. It looks like you still need to look at the 'Commits' tab on those reviews. The 'Files changed' tab seems to show changes from all of the stuff 'ahead' of the PR. But it is still easier this way I think.

File: layers/gpu_validation.cpp [hunk 1]
```diff
@@ -2034,6 +2035,21 @@ void GpuAssisted::AllocateValidationResources(const VkCommandBuffer cmd_buffer,
...
+        ReportSetupProblem(device, "Unrecognized command buffer");
+        aborted = true;
+        return;
+    }
+    const auto lv_bind_point = ConvertToLvlBindPoint(bind_point);
+    auto const &state = cb_node->lastBound[lv_bind_point];
+    const auto *pipeline_state = state.pipeline_state;
+
+    // TODO (ncesario) GPUAV not currently supported with graphics pipeline libraries
+    if (pipeline_state->IsGraphicsLibrary()) {
```
Comment (Tony-LunarG) [inline layers/gpu_validation.cpp:2049]: Is there / should there be a warning that GPU-AV isn't supported with GPLs?  Also is there a way to know at CreatePipelineLayout time if GPLs are in use and abort modification of the DSL to accommodate the debug descriptor set?  Is the same true for debug printf?

File: tests/vkrenderframework.h [hunk 2]
```diff
@@ -305,6 +305,18 @@ class VkRenderFramework : public VkTestFramework {
     // After instance and physical device creation (e.g., after InitFramework), returns true if all required extensions are
     // available, false otherwise
     bool AreRequestedExtensionsEnabled() const;
+    template <typename Container>
+    bool AreRequestedExtensionsEnabled(Container &ret) const {
```
Comment (Tony-LunarG) [inline tests/vkrenderframework.h:309]: This is great!

**Review by Tony-LunarG — COMMENTED** [review]
From what I can tell, this LGTM.  Lots of code....

**Comment by ncesario-lunarg** [comment]
> From what I can tell, this LGTM. Lots of code....

Yep, sorry for the giant PR! I tried to break this up into more manageable PRs: https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/3946#issuecomment-1087877197

File: layers/gpu_validation.cpp (hunk 1 above)
Comment (ncesario-lunarg) [reply layers/gpu_validation.cpp:2049]: > Is there / should there be a warning that GPU-AV isn't supported with GPLs?

There currently is not; this only prevents the crash. I will add a warning with a comment to remove once support is added

>  Also is there a way to know at CreatePipelineLayout time if GPLs are in use and abort modification of the DSL to accommodate the debug descriptor set?

Unfortunately, I don't think there is any way to know whether or not GPLs are going to be used at `CreatePipelineLayout` time. When you create the layout, I don't think you can then use it with GPLs or without them.

> Is the same true for debug printf?

Yep! Pretty sure I am missing a return for that. I created a test for GPUAV, but don't think I tried this for debug printf. l'll add similar code for that as well as the warning.

**Review by jeremyg-lunarg — APPROVED** [review]
LGTM pending conflict resolution. Reviewed in the 'chunked' PRs listed in the comments.

**Pull request merged**